File: nano/core_test/conflicts.cpp
#include <nano/lib/blocks.hpp>
#include <nano/node/active_elections.hpp>
#include <nano/node/election.hpp>
#include <nano/node/scheduler/component.hpp>
#include <nano/node/scheduler/priority.hpp>
#include <nano/secure/ledger.hpp>
#include <nano/test_common/chains.hpp>
#include <nano/test_common/system.hpp>
#include <nano/test_common/testutil.hpp>

#include <gtest/gtest.h>

#include <boost/variant/get.hpp>

using namespace std::chrono_literals;

TEST (conflicts, add_two)
{
	nano::test::system system{};
	auto const & node = system.add_node ();
	nano::keypair key1, key2, key3;
	auto gk = nano::dev::genesis_key;

	// create 2 new accounts, that receive 1 raw each, all blocks are force confirmed
	auto [send1, open1] = nano::test::setup_new_account (system, *node, 1, gk, key1, gk.pub, true);
	auto [send2, open2] = nano::test::setup_new_account (system, *node, 1, gk, key2, gk.pub, true);
	ASSERT_EQ (5, node->ledger.cemented_count ());

	// send 1 raw to account key3 from key1
	auto send_a = nano::state_block_builder ()
				  .account (key1.pub)
				  .previous (open1->hash ())
				  .representative (nano::dev::genesis_key.pub)
				  .balance (0)
				  .link (key3.pub)
				  .sign (key1.prv, key1.pub)
				  .work (*system.work.generate (open1->hash ()))
				  .build ();

	// send 1 raw to account key3 from key2
	auto send_b = nano::state_block_builder ()
				  .account (key2.pub)
				  .previous (open2->hash ())
				  .representative (nano::dev::genesis_key.pub)
				  .balance (0)
				  .link (key3.pub)
				  .sign (key2.prv, key2.pub)
				  .work (*system.work.generate (open2->hash ()))
				  .build ();

	// activate elections for the previous two send blocks (to account3) that we did not forcefully confirm
	ASSERT_TRUE (nano::test::process (*node, { send_a, send_b }));
	ASSERT_TRUE (nano::test::start_elections (system, *node, { send_a, send_b }));
	ASSERT_TRUE (node->active.election (send_a->qualified_root ()));
	ASSERT_TRUE (node->active.election (send_b->qualified_root ()));
	ASSERT_TIMELY_EQ (5s, node->active.size (), 2);
}

File: node/tests/tests/conflicts.rs
use rsban_core::{Amount, BlockSideband, PrivateKey, SavedBlock, UnsavedBlockLatticeBuilder};
use rsban_network::ChannelId;
use rsban_node::block_processing::BlockSource;
use std::time::Duration;
use test_helpers::{assert_timely, assert_timely_eq, start_election, start_elections, System};

#[test]
fn start_stop() {
    let mut system = System::new();
    let node1 = system.make_node();
    let mut lattice = UnsavedBlockLatticeBuilder::new();
    let key1 = PrivateKey::new();
    let send1 = lattice.genesis().send(&key1, Amount::MAX);
    node1.process(send1.clone()).unwrap();
    assert_eq!(node1.active.len(), 0);
    let election1 = start_election(&node1, &send1.hash());
    assert_eq!(node1.active.len(), 1);
    assert_eq!(election1.vote_count(), 1);
}

#[test]
fn add_existing() {
    let mut system = System::new();
    let node1 = system.make_node();

    let mut lattice = UnsavedBlockLatticeBuilder::new();
    let key1 = PrivateKey::new();

    // create a send block to send all of the nano supply to key1
    let send1 = lattice.genesis().send(&key1, Amount::MAX);

    // add the block to ledger as an unconfirmed block
    node1.process(send1.clone()).unwrap();

    // instruct the election scheduler to trigger an election for send1
    start_election(&node1, &send1.hash());

    // wait for election to be started before processing send2
    assert_timely(Duration::from_secs(5), || node1.active.active(&send1));

    let mut fork_lattice = UnsavedBlockLatticeBuilder::new();
    let key2 = PrivateKey::new();
    let send2 = fork_lattice.genesis().send(&key2, Amount::MAX);
    let send2 = SavedBlock::new(send2, BlockSideband::new_test_instance());

    // the block processor will notice that the block is a fork and it will try to publish it
    // which will update the election object
    node1
        .block_processor
        .add(send2.clone().into(), BlockSource::Live, ChannelId::LOOPBACK);

    assert!(node1.active.active(&send1));
    assert_timely(Duration::from_secs(5), || {
        node1.active.active_root(&send2.qualified_root())
    });
}

#[test]
fn add_two() {
    let mut system = System::new();
    let node = system.make_node();

    let mut lattice = UnsavedBlockLatticeBuilder::new();
    let key1 = PrivateKey::from(1);
    let key2 = PrivateKey::from(2);
    let key3 = PrivateKey::from(3);

    // create 2 new accounts, that receive 1 raw each, all blocks are force confirmed
    let send1 = lattice.genesis().send(&key1, 1);
    let send2 = lattice.genesis().send(&key2, 1);
    let open1 = lattice.account(&key1).receive(&send1);
    let open2 = lattice.account(&key2).receive(&send2);
    node.process_and_confirm_multi(&[send1, open1.clone(), send2, open2.clone()]);

    let send_a = lattice.account(&key1).send(&key3, 1);
    let send_b = lattice.account(&key2).send(&key3, 1);

    // activate elections for the previous two send blocks (to account3) that we did not forcefully confirm
    node.process(send_a.clone()).unwrap();
    node.process(send_b.clone()).unwrap();
    start_elections(&node, &[send_a.hash(), send_b.hash()], false);

    assert!(node.active.election(&send_a.qualified_root()).is_some());
    assert!(node.active.election(&send_b.qualified_root()).is_some());
    assert_timely_eq(Duration::from_secs(5), || node.active.len(), 2);
}
